src/program.rs

```diff
@@ -114,27 +118,8 @@ impl<Ext: extension::Jet> Program<Ext> {
         };
 
         let typed_program = add_witness_data(typed_program, witness_bits)?;
-
-        // Compute cached data and return
-        let mut ret = Vec::<ProgramNode<Ext>>::with_capacity(typed_program.len());
-        for (index, node) in typed_program.into_iter().enumerate() {
-            let final_node = ProgramNode {
-                index: index,
-                cmr: compute_cmr(&ret, &node.node, index),
-                extra_cells_bound: compute_extra_cells_bound(
-                    &ret,
-                    &node.node,
-                    index,
-                    node.target_ty.bit_width(),
-                ),
-                frame_count_bound: compute_frame_count_bound(&ret, &node.node, index),
-                node: node.node,
-                source_ty: node.source_ty,
-                target_ty: node.target_ty,
-            };
-            ret.push(final_node);
-        }
-        Ok(Program { nodes: ret })
+        let finalized_program = compress_and_finalize(typed_program);
```

Comment:
This should return Err if the `typed_program` is uncompressed. Not having maximal sharing is a dos vector in simplicity and would not be allowed by consensus rules. 

We should enforce in type system only fully compressed programs ever constructed in `Program`. 

We should offer programmatic utilities to help users compress a program, but while decoding programs from the wire, we should error if they are not compressed.

Reply:
We should revisit our type of safety guarantees for data structures to make sure that users cannot create invalid data structures. 

At the very least, we should do this for `Program`. This would involve removing `pub` from fields of `Program`, and only keeping necessary constructors.